lib/input/sftp.go

```diff
@@ -0,0 +1,495 @@
+package input
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/Jeffail/benthos/v3/internal/codec"
+	"io"
+	"net"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Jeffail/benthos/v3/internal/docs"
+	"github.com/Jeffail/benthos/v3/lib/input/reader"
+	"github.com/Jeffail/benthos/v3/lib/log"
+	"github.com/Jeffail/benthos/v3/lib/message"
+	"github.com/Jeffail/benthos/v3/lib/metrics"
+	"github.com/Jeffail/benthos/v3/lib/types"
+	"github.com/pkg/sftp"
+	"golang.org/x/crypto/ssh"
+)
+
+func init() {
+	var credentialsFields = docs.FieldSpecs{
+		docs.FieldCommon("username", "The username to connect to the SFTP server."),
+		docs.FieldCommon("secret", "The secret/password for the username to connect to the SFTP server."),
+	}
+
+	Constructors[TypeSFTP] = TypeSpec{
+		constructor: func(conf Config, mgr types.Manager, log log.Modular, stats metrics.Type) (Type, error) {
+			r, err := NewSFTP(conf.SFTP, log, stats)
+			if err != nil {
+				return nil, err
+			}
+			return NewAsyncReader(
+				TypeSFTP,
+				true,
+				reader.NewAsyncBundleUnacks(
+					reader.NewAsyncPreserver(r),
+				),
+				log, stats,
+			)
+		},
+		Status: docs.StatusExperimental,
+		Summary: `
+Downloads objects via an SFTP connection.`,
+		Description: `
+Downloads objects via an SFTP connection.
+## Metadata
+This input adds the following metadata fields to each message:
+` + "```" + `
+- sftp_file_path
+` + "```" + `
+You can access these metadata fields using [function interpolation](/docs/configuration/interpolation#metadata).`,
+
+		FieldSpecs: docs.FieldSpecs{
+			docs.FieldCommon(
+				"server",
+				"The server to connect to that has the target files.",
+			),
+			docs.FieldCommon(
+				"port",
+				"The port to connect to on the server.",
+			),
+			docs.FieldCommon(
+				"credentials",
+				"The credentials to use to log into the server.",
+			).WithChildren(credentialsFields...),
+			docs.FieldCommon(
+				"filename",
+				"The name of the file to pull messages from. If not provided, all the files in the path will be processed.",
+			),
+			docs.FieldCommon(
+				"path",
+				"The path of the directory or file that it will process.",
+			),
+			docs.FieldCommon(
+				"max_connection_attempts",
+				"How many times it will try to connect to the server before exiting with an error.",
+			),
+			docs.FieldAdvanced(
+				"retry_sleep_duration",
+				"How long it will sleep after failing to connect to the server before trying again, defaults to 5s if not provided.",
+				"10s", "5m",
+			),
+			codec.ReaderDocs,
+			docs.FieldAdvanced("delete_objects", "Whether to delete files from the server once they are processed."),
+		},
+		Categories: []Category{
+			CategoryServices,
+			CategoryNetwork,
+		},
+	}
+}
+
+//------------------------------------------------------------------------------
+
+// SFTPConfig contains configuration fields for the SFTP input type.
+type SFTPConfig struct {
+	Server                string          `json:"server" yaml:"server"`
```

Comment:
In most places we have a single field called `address` for host:port combos, it'd be good to follow that here for consistency, you can then use `url.Parse` to pull out the individual values.